ddtrace/internal/service.py

```diff
@@ -56,5 +76,87 @@ def stop(self):
         self.status = ServiceStatus.STOPPED
 
     @staticmethod
-    def join(timeout=None):
+    def join(
+        timeout=None,  # type: typing.Optional[float]
+    ):
+        # type: (...) -> None
         """Join the service once stopped."""
+
+    def copy(self):
+        # type: () -> Service
+        return attr.evolve(self)
+
+
+class ForksafeService(object):
+    """A manager that can handle the lifecycle of a service in a process.
+
+    This allows to handle a service that should keep running when, e.g., the Python process forks.
+    """
+
+    SERVICE_CLASS = None  # type: typing.ClassVar[typing.Type[Service]]
+
+    def __init__(
+        self,
+        *args,  # type: typing.Any
+        **kwargs  # type: typing.Any
+    ):
+        # type: (...) -> None
+        super(ForksafeService, self).__init__()
+        self._service = self.SERVICE_CLASS(*args, **kwargs)  # type: ignore[call-arg]
+
+    def start(self, *args, **kwargs):
+        # type: (...) -> None
+        """Start the service.
+
+        :param stop_on_exit: Whether to stop the service and wait for shutdown on exit.
+        :param start_in_children: Whether to start the service in child processes.
+        """
+        stop_on_exit = kwargs.pop("stop_on_exit", True)
+        start_in_children = kwargs.pop("start_in_children", True)
+
+        if start_in_children:
+            try:
+                uwsgi.check_uwsgi(self._restart_on_fork, atexit=self.stop if stop_on_exit else None)
```

Comment:
This would make it so that uwsgi child processes restart twice as you are registering both post fork hooks with uwsgi and with `os.register_at_fork` which will be called once `PyOS_AfterFork[_Child]` is called.

Reply:
uwsgi calls the `fork()` system calls, not Python's fork. So those hooks are not called.
Otherwise, we wouldn't need a custom uwsgi code at all. 🤷‍♂️

Reply:
After calling the system `fork()`, uwsgi does end up calling PyOS_AfterFork_Child which in turn will call the hooks that had been registered:

* https://github.com/unbit/uwsgi/search?q=PyOS_AfterFork
* https://github.com/python/cpython/blob/95e4431804587a0c9d464bb7b3d5f3057bbeaccd/Modules/posixmodule.c#L588
* https://github.com/python/cpython/blob/95e4431804587a0c9d464bb7b3d5f3057bbeaccd/Modules/posixmodule.c#L6566-L6595

Reply:
The difference in uWSGI is that by default `PyOS_AfterFork_Child` is not called early on in the child process after it is forked. It can be called earlier if the `py-call-osafterfork` option is set.

* https://github.com/unbit/uwsgi/blob/d58a832c81c2c96ae0f6e72614e1cc47f4b5d332/plugins/python/python_plugin.c#L435-L442
* https://github.com/unbit/uwsgi/blob/d58a832c81c2c96ae0f6e72614e1cc47f4b5d332/plugins/python/python_plugin.c#L2048-L2053

Reply:
`PyOS_AfterFork_Child` is actually never called unless you enable `py-call-osafterfork` IIUC. So I think that code is still right, *except* if `py-call-osafterfork` is enabled. 😞 

That might be something we need to detect and check, even in the current `ddtrace.internal.uwsgi` module 😢

Reply:
Sorry, I misunderstood how uwsgi was doing things. `PyOS_AfterFork_Child` is called in [`uwsgi_python_post_fork`](https://github.com/unbit/uwsgi/blob/d58a832c81c2c96ae0f6e72614e1cc47f4b5d332/plugins/python/python_plugin.c#L435-L442) only if `py-call-osafterfork` is enabled. If it is disabled `PyOS_AfterFork_Child` it can still be called in [`uwsgi_python_worker`](https://github.com/unbit/uwsgi/blob/d58a832c81c2c96ae0f6e72614e1cc47f4b5d332/plugins/python/python_plugin.c#L2048-L2053) but only if you are using [worker overrides](https://uwsgi-docs.readthedocs.io/en/latest/WorkerOverride.html).

Going back to the initial comment, it would still seem like we don't want to register hooks in uwsgi and in Python if we know we are running within uwsgi. Whether or not the user has enabled `py-call-osafterfork` would not matter here as the uwsgi post fork hooks would be [called](https://github.com/unbit/uwsgi/blob/d58a832c81c2c96ae0f6e72614e1cc47f4b5d332/plugins/python/python_plugin.c#L446-L454) regardless.

As long as a uwsgi user is not using `ddtrace-run`, they could still start such `ForksafeServices` in the master process, register hooks with `ddtrace.internal.uwsgi`, and have those services restarted in the child worker processes.

Reply:
> Going back to the initial comment, it would still seem like we don't want to register hooks in uwsgi and in Python if we know we are running within uwsgi. Whether or not the user has enabled py-call-osafterfork would not matter here as the uwsgi post fork hooks would be called regardless.

This code does not register after fork hook in the master process, it only does in the children process. Which is correct, since it's possible for a uwsgi child process to call `os.fork`.

If the user enables `py-call-osafterfork` then the problem also does not arise: the `start` method does not register `forksafe` hook in the master process, it just returns right away and wait for the uwsgi postfork hook to be triggered – so we're still fine.

WDYT?

Reply:
Ok, sorry it took me a few back and forths to understand what's happening in the case of uwsgi. The restart will be registered as uwsgi post fork hook in the master but not with forksafe, in the child the restart will be registered with forksafe so if a worker calls `os.fork` then it'll sure to be started. 

Finally, `py-call-osafterfork` does not matter to this functionality as these restarts are handled in the master process by uwsgi's own post fork hooks.

Reply:
Exactly @majorgreys!

… so why don't I have a ✅  ? 😁